picom-next: init at vNext

GKasparov opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Dual_Kawase blur isn't supported on the default picom


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

This PR adds the tryone fork of picom, which supports dual_kawase blur method.
Source: https://github.com/tryone144/picom/tree/feature/dual_kawase

GKasparov committed:
picom-tryone: init at unstable-2021-10-31

GKasparov committed:
Create default.nix

GKasparov committed:
Adjust sources

GKasparov committed:
Add more info into description

GKasparov committed:
Remove final trailing whitespace

GKasparov committed:
Update default.nix

GKasparov committed:
Remove Line 105 trailing whitespace

bobby285271 commented:
Please add the package to `pkgs/top-level/all-packages.nix` and follow [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md) when writing commit messages.

altgk committed:
picom-tryone: added to all-packages.nix

Essential change to all-packages.nix to add entry for picom-tryone

GKasparov committed:
picom-tryone: fixed sha256

The last digit was wrong (I altered it manually in testing). Adjusted it to the correct sha256

GKasparov committed:
picom-tryone: Added GKasparov as a maintainer

I'll be housekeeping this package.

GKasparov committed:
picom-tryone: Set fetchSubmodules to false

The .gitmodules is empty no need to fetch

GKasparov committed:
Remove Picom-Tryone package. Add picom-next

GKasparov commented:
> An example on how to override the version directly from `pkgs/top-level/all-packages.nix` (can't link directly to GitHub link since this file is too big to GitHub to render it):
> 
> ```nix
>   libgit2_0_27 = libgit2.overrideAttrs (oldAttrs: rec {
>     version = "0.27.10";
>     src = fetchFromGitHub {
>       owner = "libgit2";
>       repo = "libgit2";
>       rev = "v${version}";
>       sha256 = "09jz2fzv0zl5058s0g1cpnw87a2rgg8wnjwlygi18i2n9nn6m0ad";
>     };
>   });
> ```

I've changed it to your suggestion, adding picom-next overrided with the next branch. Another way could be using the vNext tag rather than current committs, but I don't see the benefit in that

thiagokokada commented:
> > An example on how to override the version directly from `pkgs/top-level/all-packages.nix` (can't link directly to GitHub link since this file is too big to GitHub to render it):
> > ```nix
> >   libgit2_0_27 = libgit2.overrideAttrs (oldAttrs: rec {
> >     version = "0.27.10";
> >     src = fetchFromGitHub {
> >       owner = "libgit2";
> >       repo = "libgit2";
> >       rev = "v${version}";
> >       sha256 = "09jz2fzv0zl5058s0g1cpnw87a2rgg8wnjwlygi18i2n9nn6m0ad";
> >     };
> >   });
> > ```
> 
> I've changed it to your suggestion, adding picom-next overrided with the next branch. Another way could be using the vNext tag rather than current committs, but I don't see the benefit in that

`vNext` tag AFAIK is always changing, so it is better to pin a commit like you did.

thiagokokada commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28861,6 +28861,17 @@ with pkgs;
+      src = fetchFromGitHub {
+      owner = "yshui";
+      repo = "picom";
+      rev = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
```
This doesn't look like a valid commit.

GKasparov replied:
Updated (I accidentally pasted the wrong string.

thiagokokada replied:
Checks: https://github.com/yshui/picom/commit/fade045eadf171d2c732820d6ebde7d1943a1397

thiagokokada commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28861,6 +28861,17 @@ with pkgs;
+      rev = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
+      sha256 = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
+  };
+    maintainers = with maintainers; [ GKasparov ];
```
```suggestion
    maintainers = with maintainers; oldAttrs.maintainers ++ [ GKasparov ];
```

thiagokokada commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28861,6 +28861,17 @@ with pkgs;
   picom = callPackage ../applications/window-managers/picom {};
 
+  picom-next = picom.overrideAttrs (oldAttrs: rec {
+    version="vNext";
```
```suggestion
      version = "vNext";
```

thiagokokada replied:
Not sure if we want it to have a proper version instead.

GKasparov replied:
If so? which version
vNext indicates its the updated next branch. Idk either

thiagokokada replied:
Not sure too, you can leave it as-is for now. Maybe some other reviewer has a better idea.

GKasparov committed:
picom-next: fix maintainers, rev and whitespace

thiagokokada commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28861,6 +28861,17 @@ with pkgs;
+      rev = "fade045eadf171d2c732820d6ebde7d1943a1397";
+      sha256 = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
+  };
+    maintainers = with maintainers; oldAttrs.maintainers ++ [ GKasparov ];
```
Sorry :disappointed:.

```suggestion
    meta.maintainers = with maintainers; oldAttrs.maintainers ++ [ GKasparov ];
```

GKasparov replied:
Nws, my fault

GKasparov committed:
picom-next: fix maintainers path

thiagokokada commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28869,7 +28869,7 @@ with pkgs;
       sha256 = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
   };
-    maintainers = with maintainers; oldAttrs.maintainers ++ [ GKasparov ];
+    meta.maintainers = with maintainers; oldAttrs.maintainers ++ [ GKasparov ];
```
```suggestion
    meta.maintainers = with maintainers; oldAttrs.meta.maintainers ++ [ GKasparov ];
```

thiagokokada replied:
This was where I said sorry because this was my mistake.

thiagokokada replied:
CC @GKasparov .

SuperSandro2000 commented:
I think this is more something for a local overlay and does not need to be included in nixpkgs.

SuperSandro2000 commented:
Also this should be in an extra file rather than in top-level.

thiagokokada commented:
> I think this is more something for a local overlay and does not need to be included in nixpkgs.

Maybe. But I also think that `picom-next` has many features that some people wanted and it seems that the next stable release will still take sometime to happen. We can always add an alias `picom-next = picom` once the stable release happens.

See also: https://github.com/NixOS/nixpkgs/pull/142171#issuecomment-946837071

> Also this should be in an extra file rather than in top-level.

Any technical reason for this considering that there shouldn't be any differences in the derivation itself and also there are other examples of `overrideAtrrs` inside `all-packages.nix` :thinking: ?

GKasparov commented:
> I think this is more something for a local overlay and does not need to be included in nixpkgs.

Like thiago, I know there are ways that users locally can install picom from the next branch but the dual_kawase blur feature and the patches in the next branch are really significant and the features and wanted, so why not automate the process for people seeking it by adding the package?

GKasparov committed:
picom-next: Add correct package name

jonringer commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28861,6 +28861,18 @@ with pkgs;
 
+  picom-next = picom.overrideAttrs (oldAttrs: rec {
+  pname = "picom-next";
+    version = "vNext";
```
```suggestion
    version = "unstable-YYYY-MM-DD";
```

in accordance with https://nixos.org/manual/nixpkgs/stable/#sec-package-naming

GKasparov replied:
Resolved as indentation was fixed after moving the logic to it's own file

GKasparov committed:
picom-next: correct version string

GKasparov committed:
picom-next: changed to different file

GKasparov committed:
picom-next: add picom-next.nix

GKasparov committed:
picom-next: whitespace in picom-next.nix

GKasparov committed:
picom-next: add picom reference

GKasparov committed:
Merge remote-tracking branch 'origin/master'

thiagokokada commented on pkgs/applications/window-managers/picom/picom-next.nix:
```diff
@@ -0,0 +1,13 @@
+      rev = "fade045eadf171d2c732820d6ebde7d1943a1397";
+      sha256 = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
+      };
+    meta.maintainers = with maintainers; oldAttrs.maintainers ++ [ GKasparov ];
```
```suggestion
    meta.maintainers = with maintainers; oldAttrs.meta.maintainers ++ [ GKasparov ];
```

GKasparov replied:
Committed.

GKasparov committed:
picom-next: correct oldAttrs maintainers ref.

GKasparov committed:
picom-next: removed unnecessary fetchFromGithub

thiagokokada commented on pkgs/applications/window-managers/picom/picom-next.nix:
```diff
@@ -0,0 +1,13 @@
+{ picom }:
+
+picom.overrideAttrs (oldAttrs: rec {
+  pname = "picom-next";
+    version = "unstable-2021-11-01";
+    src = fetchFromGitHub {
+      owner = "yshui";
+      repo = "picom";
+      rev = "fade045eadf171d2c732820d6ebde7d1943a1397";
+      sha256 = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
+      };
+    meta.maintainers = with maintainers; oldAttrs.meta.maintainers ++ [ GKasparov ];
+  })
```
```suggestion
{ picom, lib, fetchFromGitHub }:

picom.overrideAttrs (oldAttrs: rec {
  pname = "picom-next";
  version = "unstable-2021-11-01";
  src = fetchFromGitHub {
    owner = "yshui";
    repo = "picom";
    rev = "fade045eadf171d2c732820d6ebde7d1943a1397";
    sha256 = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
  };
  meta.maintainers = with lib.maintainers; oldAttrs.meta.maintainers ++ [ GKasparov ];
})
```

GKasparov replied:
Added lib import, changed to lib.maintainers and fixed indents

GKasparov committed:
picom-next: once again, fixed maintainers

GKasparov committed:
picom-next: addid lib, fetchFromGithub imports

GKasparov committed:
picom-next: addid lib, fetchFromGithub imports, fixed indents

GKasparov committed:
picom-next: GitHub rather than Github

thiagokokada commented on pkgs/applications/window-managers/picom/picom-next.nix:
```diff
@@ -0,0 +1,13 @@
+    rev = "fade045eadf171d2c732820d6ebde7d1943a1397";
+    sha256 = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
+  };
+  meta.maintainers = with lib.maintainers; oldAttrs.meta.maintainers ++ [ GKasparov ];
```
You need to add yourself on `maintainers/maintainers-list.nix` file.

thiagokokada reviewed (commented):
Code looks fine now.

Please add yourself to `maintainers/maintainers-list.nix` file (at alphabetical order) and squash your commits, and you should be good to go.

thiagokokada commented:
> Code looks fine now.
> 
> Please add yourself to `maintainers/maintainers-list.nix` file (at alphabetical order) and squash your commits, and you should be good to go.

The final PR should have at most two commits:

- `maintainers: add GKasparov`
- `picom-next: init at unstable-2021-11-01`

thiagokokada commented on pkgs/applications/window-managers/picom/picom-next.nix:
```diff
@@ -0,0 +1,13 @@
+
+picom.overrideAttrs (oldAttrs: rec {
+  pname = "picom-next";
+  version = "unstable-2021-11-01";
```
```suggestion
  version = "unstable-2021-10-31";
```

Date must be the date of fetched commit (not the date of the PR).

GKasparov replied:
Sorry im in a different timezone so i got the date wrong

GKasparov committed:
picom-next: add GKasparov to maintainer-list

GKasparov committed:
picom-next: add GKasparov to maintainer-list

GKasparov committed:
picom-next: Change version date to commit

thiagokokada reviewed (commented):
Result of `nixpkgs-review pr 142171` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>picom-next</li>
  </ul>
</details>

Works now, please squash the commits as instructed and you're good to go.

GKasparov commented:
> Result of `nixpkgs-review pr 142171` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 1 package built:
> 
>     * picom-next
> 
> 
> Works now, please squash the commits as instructed and you're good to go.

How can I squash without write access. The drop down is usually next to the merge button

thiagokokada commented:
> > Result of `nixpkgs-review pr 142171` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> > 1 package built:
> > ```
> > * picom-next
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Works now, please squash the commits as instructed and you're good to go.
> 
> How can I squash without write access. The drop down is usually next to the merge button

You need to squash manually in your fork and do a force push. Something like:

```
$ git remote add upstream https://github.com/NixOS/nixpkgs/
$ git fetch upstream
$ git rebase -i upstream/master # this will open your text editor so you can manually rebase your changes
$ git push --force
```

GKasparov committed:
Merge pull request #1 from GKasparov/picom-next

Picom next

GKasparov commented:
Hey, I've remade  the PR cleaner and with the proper commit messages. The way my branches were setup was extremely messy so please close this and merge the new one: https://github.com/NixOS/nixpkgs/pull/144696
